arch/RISCV/RISCVInstPrinter.c

```diff
@@ -0,0 +1,355 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This class prints an RISC-V MCInst to a .s file.
+//
+//===----------------------------------------------------------------------===//
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <capstone/platform.h>
+
+#include "RISCVMapping.h"
+#include "RISCVInstPrinter.h"
+
+#define GET_SUBTARGETINFO_ENUM
+#include "RISCVGenSubtargetInfo.inc"
+
+#define GET_INSTRINFO_ENUM
+#include "RISCVGenInstrInfo.inc"
+
+#define GET_REGINFO_ENUM
+#include "RISCVGenRegisterInfo.inc"
+
+#define CONCAT(a, b) CONCAT_(a, b)
+#define CONCAT_(a, b) a##_##b
+
+#define DEBUG_TYPE "asm-printer"
+
+// Include the auto-generated portion of the assembly writer.
+#define PRINT_ALIAS_INSTR
+#include "RISCVGenAsmWriter.inc"
+
+// Print architectural register names rather than the ABI names (such as x2
+// instead of sp).
+// TODO: Make RISCVInstPrinter::getRegisterName non-static so that this can a
+// member.
+static bool ArchRegNames;
+
+void printRegName(SStream *O, MCRegister Reg)
+{
+	SStream_concat0(markup(O, Markup_Register), getRegisterName(Reg));
+}
+
+static inline void printOperand(MCInst *MI, unsigned OpNo, SStream *O,
+				const char *Modifier)
+{
+	RISCV_add_cs_detail_0(MI, RISCV_OP_GROUP_Operand, OpNo);
+	CS_ASSERT((Modifier == nullptr || Modifier[0] == 0) &&
+		  "No modifiers supported");
+	MCOperand *MO = MCInst_getOperand(MI, (OpNo));
+
+	if (MCOperand_isReg(MO)) {
+		printRegName(O, MCOperand_getReg(MO));
+		return;
+	}
+
+	if (MCOperand_isImm(MO)) {
+		SStream_concat0(markup(O, Markup_Immediate),
+				formatImm(MCOperand_getImm(MO)));
+		return;
+	}
+
+	CS_ASSERT(MCOperand_isExpr(MO) &&
+		  "Unknown operand kind in printOperand");
+	MCOperand_getExpr(MO)->print(O, &MAI);
+}
+
+static inline void printBranchOperand(MCInst *MI, uint64_t Address,
+				      unsigned OpNo, SStream *O)
+{
+	RISCV_add_cs_detail_0(MI, RISCV_OP_GROUP_BranchOperand, OpNo);
+	MCOperand *MO = MCInst_getOperand(MI, (OpNo));
+	if (!MCOperand_isImm(MO))
+		return printOperand(MI, OpNo, O);
+
+	if (PrintBranchImmAsAddress) {
+		uint64_t Target = Address + MCOperand_getImm(MO);
+		if (!STI.hasFeature(RISCV_Feature64Bit))
+			Target &= 0xffffffff;
+		SStream_concat0(markup(O, Markup_Target), formatHex(Target));
+	} else {
+		SStream_concat0(markup(O, Markup_Target),
+				formatImm(MCOperand_getImm(MO)));
+	}
+}
+
+static inline void printCSRSystemRegister(MCInst *MI, unsigned OpNo, SStream *O)
+{
+	RISCV_add_cs_detail_0(MI, RISCV_OP_GROUP_CSRSystemRegister, OpNo);
+	unsigned Imm = MCOperand_getImm(MCInst_getOperand(MI, (OpNo)));
+	auto SysReg = RISCVSysReg_lookupSysRegByEncoding(Imm);
+	if (SysReg && SysReg->haveRequiredFeatures(STI.getFeatureBits()))
+		SStream_concat0(markup(O, Markup_Register), SysReg->Name);
+	else
+		SStream_concat0(markup(O, Markup_Register), formatImm(Imm));
```

Comment:
Yeah, better get rid of these `formatImm` calls. Otherwise we have two places to take care of formatting and everything gets less consistent.